src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+            (_, prefix, suffix) = foldr' sepWith (0, [], []) sfp
+            toShortPath = toShort . S8.intercalate pathSeparatorS
+        in (toShortPath prefix, toShortPath suffix)
+
+packHeader :: MonadThrow m => Header -> m S.ByteString
+packHeader header = do
+    (left, right) <- packHeaderNoChecksum header
+    let sumsl :: SL.ByteString -> Int
+        sumsl = SL.foldl' (\ !acc !v -> acc + fromIntegral v) 0
+    encChecksum <- encodeOctal 7 $ sumsl left + 32 * 8 + sumsl right
+    return $
+        SL.toStrict $
+        toLazyByteString $ lazyByteString left <> encChecksum <> word8 0 <> lazyByteString right
+
+
+packHeaderNoChecksum :: MonadThrow m => Header -> m (SL.ByteString, SL.ByteString)
+packHeaderNoChecksum Header {..} = do
+    let CTime headerTime' = headerTime
+    hNameSuffix <- encodeShort 100 headerFileNameSuffix
+    hFileMode <- encodeOctal 7 headerFileMode
+    hOwnerId <- encodeOctal 7 headerOwnerId
+    hGroupId <- encodeOctal 7 headerGroupId
+    hPayloadSize <- encodeOctal 11 headerPayloadSize
+    hTime <- encodeOctal 11 headerTime'
+    hLinkName <- encodeShort 100 headerLinkName
+    hMagicVersion <- encodeShort 8 headerMagicVersion
+    hOwnerName <- encodeShort 32 headerOwnerName
+    hGroupName <- encodeShort 32 headerGroupName
+    hDevMajor <- encodeDevice headerDeviceMajor
+    hDevMinor <- encodeDevice headerDeviceMinor
+    hNamePrefix <- encodeShort 155 headerFileNamePrefix
+    return
+        ( toLazyByteString $
+          hNameSuffix <>
+          hFileMode <> word8 0 <>
+          hOwnerId <> word8 0 <>
+          hGroupId <> word8 0 <>
+          hPayloadSize <> word8 0 <>
+          hTime <> word8 0
+        , toLazyByteString $
+          word8 headerLinkIndicator <>
+          hLinkName <>
+          hMagicVersion <>
+          hOwnerName <>
+          hGroupName <>
+          hDevMajor <> word8 0 <>
+          hDevMinor <> word8 0 <>
+          hNamePrefix <>
+          byteString (S.replicate 12 0)
+        )
+  where
+    encodeDevice 0     = return $ byteString $ S.replicate 7 0
+    encodeDevice devid = encodeOctal 7 devid
+
+
+-- | Encode a `ShortByteString` with an exact length, NUL terminating if it is
+-- shorter, but throwing `TarCreationError` if it is longer.
+encodeShort :: MonadThrow m => Int -> ShortByteString -> m Builder
+encodeShort !len !sbs
+    | lenShort <= len = return $ byteString $ fst $ S.unfoldrN len maybeShort 0
+    | otherwise =
+        throwM $
+        TarCreationError $ "Can't fit '" ++ S8.unpack (fromShort sbs) ++ "' into the tar header"
+  where
+    lenShort = SS.length sbs
+    maybeShort !i
+        | i < lenShort = Just (SS.index sbs i, i + 1)
+        | otherwise = Just (0, i + 1)
+
+
+-- | Encode a number in 8base padded with zeros. Throws `TarCreationError` when overflows.
+encodeOctal :: (Show a, Integral a, MonadThrow m) => Int -> a -> m Builder
+encodeOctal !len !val = do
+    enc <- case S.unfoldrN len toOctal val of
```

Comment:
It would be more performant to avoid construction of the intermediate `ByteString` and instead construct a `Builder` directly.

Reply:
You got me thinking on this one, so I decided to experiment a bit with alternative implementation using `Builder` directly: https://gist.github.com/lehins/0e15d8bd2e6cd660d835cd7b0cd90c6c
Result is that optimization is minuscule and was not worth the trouble, but was fun to play around with anyways. :) Another confirmation that padding with `S.replicate` is the way to go.